### ISSUE TITLE
mod_sftp in 1.3.7 still has issues with OpenSSL 3.x

### ISSUE THREAD

**Issue by beckerr-rzht** [proftpd/proftpd#1470]
### What I Did

`mod_sftp` was configured and running on Ubuntu 20.04 LTS.
Upgraded running installation to Ubuntu 22.04 LTS.

### What I Expected/Wanted

I expect `mod_sftp` still serves data, but it didn't.
Error: `error encrypting aes128-ctr data for client`

Does anyone have any idea where the cause of the problem could be?
Openssl 3?
`OpenSSL 3.0.2 15 Mar 2022 (Library: OpenSSL 3.0.2 15 Mar 2022)`

Logs:
```
2022-06-10 10:39:20,225 mod_sftp/1.0.1[656719]: received client version 'SSH-2.0-OpenSSH_8.9p1 Ubuntu-3'
2022-06-10 10:39:20,225 mod_sftp/1.0.1[656719]: handling connection from SSH2 client 'OpenSSH_8.9p1'
2022-06-10 10:39:20,251 mod_sftp/1.0.1[656719]:  + Session key exchange: curve25519-sha256
2022-06-10 10:39:20,251 mod_sftp/1.0.1[656719]:  + Session server hostkey: rsa-sha2-512
2022-06-10 10:39:20,251 mod_sftp/1.0.1[656719]:  + Session client-to-server encryption: aes128-ctr
2022-06-10 10:39:20,251 mod_sftp/1.0.1[656719]:  + Session server-to-client encryption: aes128-ctr
2022-06-10 10:39:20,251 mod_sftp/1.0.1[656719]:  + Session client-to-server MAC: umac-64@openssh.com
2022-06-10 10:39:20,251 mod_sftp/1.0.1[656719]:  + Session server-to-client MAC: umac-64@openssh.com
2022-06-10 10:39:20,251 mod_sftp/1.0.1[656719]:  + Session client-to-server compression: none
2022-06-10 10:39:20,251 mod_sftp/1.0.1[656719]:  + Session server-to-client compression: none
2022-06-10 10:39:20,302 mod_sftp/1.0.1[656719]: error encrypting aes128-ctr data for client: (unknown)
2022-06-10 10:39:20,302 mod_sftp/1.0.1[656719]: disconnecting 10.249.140.7 (Application disconnected)
2022-06-10 10:39:20,302 mod_sftp/1.0.1[656719]: error encrypting aes128-ctr data for client: (unknown)
```

### ProFTPD Version and Configuration

```bash
proftpd -V
Compile-time Settings:
  Version: 1.3.7c (maint)
  Platform: LINUX [Linux 5.15.0-33-generic x86_64]
  Built: Fri Dec 3 2021 13:35:22 UTC
  Built With:
    configure  '--infodir=/share/info' '--disable-option-checking' '--disable-silent-rules' '--libdir=/lib/x86_64-linux-gnu' '--disable-dependency-tracking' '--prefix=/usr' '--with-pkgconfig=lib/pkgconfig' '--with-includes=/usr/include/postgresql:/usr/include/mysql' '--mandir=/usr/share/man' '--sysconfdir=/etc/proftpd' '--localstatedir=/run' '--libexecdir=/usr/lib/proftpd' '--enable-sendfile' '--enable-facl' '--enable-dso' '--enable-autoshadow' '--enable-ctrls' '--enable-openssl' '--enable-ipv6' '--enable-nls' '--enable-memcache' '--with-lastlog=/var/log/lastlog' '--enable-pcre' '--disable-strip' '--enable-redis' '--build' 'x86_64-linux-gnu' '--with-shared=mod_unique_id:mod_site_misc:mod_load:mod_ban:mod_quotatab:mod_sql:mod_sql_mysql:mod_sql_postgres:mod_sql_sqlite:mod_sql_odbc:mod_dynmasq:mod_quotatab_sql:mod_ldap:mod_quotatab_ldap:mod_ratio:mod_tls:mod_rewrite:mod_radius:mod_wrap:mod_wrap2:mod_wrap2_file:mod_wrap2_sql:mod_quotatab_file:mod_quotatab_radius:mod_facl:mod_ctrls_admin:mod_copy:mod_deflate:mod_ifversion:mod_geoip:mod_exec:mod_sftp:mod_sftp_pam:mod_sftp_sql:mod_shaper:mod_sql_passwd:mod_ifsession:mod_auth_otp:mod_tls_redis:mod_wrap2_redis:mod_redis:mod_memcache:mod_tls_memcache:mod_readme:mod_snmp:mod_digest:mod_ident:mod_log_forensic:mod_qos:mod_statcache:mod_tls_fscache:mod_tls_shmcache:mod_dnsbl' 'build_alias=x86_64-linux-gnu' 'CFLAGS=-g -O2 -ffile-prefix-map=/build/proftpd-dfsg-iPbrOs/proftpd-dfsg-1.3.7c+dfsg=. -flto=auto -ffat-lto-objects -fstack-protector-strong -Wformat -Werror=format-security' 'LDFLAGS=-Wl,-Bsymbolic-functions -flto=auto -Wl,-z,relro -Wl,-z,now' 'CPPFLAGS=-Wdate-time -D_FORTIFY_SOURCE=2' 'CXXFLAGS=-g -O2 -ffile-prefix-map=/build/proftpd-dfsg-iPbrOs/proftpd-dfsg-1.3.7c+dfsg=. -flto=auto -ffat-lto-objects -fstack-protector-strong -Wformat -Werror=format-security'

  CFLAGS: -g2 -g -O2 -ffile-prefix-map=/build/proftpd-dfsg-iPbrOs/proftpd-dfsg-1.3.7c+dfsg=. -flto=auto -ffat-lto-objects -fstack-protector-strong -Wformat -Werror=format-security -Wall -fno-omit-frame-pointer -fno-strict-aliasing -Werror=implicit-function-declaration
  LDFLAGS: -L$(top_srcdir)/lib -L$(top_builddir)/lib -Wl,-Bsymbolic-functions -flto=auto -Wl,-z,relro -Wl,-z,now -rdynamic  -L/usr/lib/x86_64-linux-gnu -L/usr/lib/x86_64-linux-gnu
  LIBS: -lacl  -lpcreposix -lpcre -lssl -lcrypto -lsodium -lcap  -lpam -lsupp -lattr -lnsl -lresolv -lresolv -lcrypt -lhiredis -lmemcachedutil -lmemcached  -pthread

  Files:
    Configuration File:
      /etc/proftpd/proftpd.conf
    Pid File:
      /run/proftpd.pid
    Scoreboard File:
      /run/proftpd.scoreboard
    Header Directory:
      /usr/include/proftpd
    Shared Module Directory:
      /usr/lib/proftpd

  Info:
    + Max supported UID: 4294967295
    + Max supported GID: 4294967295

  Features:
    + Autoshadow support
    + Controls support
    + curses support
    - Developer support
    + DSO support
    + IPv6 support
    + Largefile support
    + Lastlog support
    + Memcache support
    + ncursesw support
    + NLS support
    + OpenSSL support (OpenSSL 3.0.0 7 sep 2021)
    + PCRE support
    + POSIX ACL support
    + Redis support
    + Sendfile support
    + Shadow file support
    + Sodium support
    + Trace support
    + xattr support

  Tunable Options:
    PR_TUNABLE_BUFFER_SIZE = 1024
    PR_TUNABLE_DEFAULT_RCVBUFSZ = 8192
    PR_TUNABLE_DEFAULT_SNDBUFSZ = 8192
    PR_TUNABLE_ENV_MAX = 2048
    PR_TUNABLE_GLOBBING_MAX_MATCHES = 100000
    PR_TUNABLE_GLOBBING_MAX_RECURSION = 8
    PR_TUNABLE_HASH_TABLE_SIZE = 40
    PR_TUNABLE_LOGIN_MAX = 256
    PR_TUNABLE_NEW_POOL_SIZE = 512
    PR_TUNABLE_PATH_MAX = 4096
    PR_TUNABLE_SCOREBOARD_BUFFER_SIZE = 80
    PR_TUNABLE_SCOREBOARD_SCRUB_TIMER = 30
    PR_TUNABLE_SELECT_TIMEOUT = 30
    PR_TUNABLE_TIMEOUTIDENT = 10
    PR_TUNABLE_TIMEOUTIDLE = 600
    PR_TUNABLE_TIMEOUTLINGER = 10
    PR_TUNABLE_TIMEOUTLOGIN = 300
    PR_TUNABLE_TIMEOUTNOXFER = 300
    PR_TUNABLE_TIMEOUTSTALLED = 3600
    PR_TUNABLE_XFER_SCOREBOARD_UPDATES = 10
```
`mod_sftp` specific configurations:
```
        <IfModule mod_sftp.c>

                SFTPEngine on
                SFTPLog /var/log/proftpd/sftp.log
                SFTPPAMEngine off

                # Configure both the RSA and DSA host keys, using the same host key
                # files that OpenSSH uses.
                SFTPHostKey /etc/ssh/ssh_host_rsa_key
                SFTPHostKey /etc/ssh/ssh_host_dsa_key

                # Enable compression
                SFTPCompression delayed

                # Allow the same number of authentication attempts as OpenSSH.
                #
                # It is recommended that you explicitly configure MaxLoginAttempts
                # for your SSH2/SFTP instance to be higher than the normal
                # MaxLoginAttempts value for FTP, as there are more ways to authenticate
                # using SSH2.
                MaxLoginAttempts 6

                SFTPClientMatch .*ClientSftp sftpProtocolVersion 3

                # Remove the CBC ciphers from the ciphers we advertise
                SFTPCiphers aes256-ctr aes192-ctr aes128-ctr blowfish-ctr arcfour256 arcfour128 3des-ctr

                # Question: When I create a directory via SFTP, in a directory with the SGID bit set, the created
                # directory does not have the SGID bit set as expected. How do I fix this?
                # Answer: Preserving the SGID bit for a created directory usually require root privs; the mod_sftp
                # module automatically drops all root privs. So to get the directory SGID bit functionality, you will
                # need to use the following in your mod_sftp <VirtualHost> section:
                RootRevoke off
        </IfModule>
```

**Comment by Castaglia** [comment]
I believe this is the same as Issue #1448 .

**Comment by MrMacvos** [comment]
I encounter the same SFTP encrypting error on a Fedora 36 server, with ProFTPd version 1.3.7d-1.fc36.
I have read in #1448 that it has been fixed for v1.3.8 ? How can I update to v1.3.8 ?

**Comment by tnolte1** [comment]
I had exactly the same error: "error encrypting aes128-ctr data for client: (unknown)". Neither 1.3.7d nor 1.3.8rc3 worked. But checking out and compiling the master branch worked just fine.

**Comment by gojun077** [comment]
Hi ~ I'm running 1.3.7d on Fedora 36 and I created a src rpm patch using commit [8aa39b2](https://github.com/proftpd/proftpd/commit/8aa39b27d8fd6ada556b51c4547a504956474078) which has changes to 7 files under `../contrib/mod_sftp/` - `cipher.c`, `configure`, `configure.in`, `crypto.c`, `keys.c`, `mod_sftp.c`,  and `mod_sftp.h.in`. I chose this particular commit because the message reads "Backporting the mod_sftp/OpenSSL 3.x fixes to the 1.3.7 branch".

After creating custom patched rpm's from this commit, I tested them in the same environment that is running an unpatched `proftpd 1.3.7d` but I get the same aes128-ctr error with SFTP connections.

**Question**: Should I use a later commit when creating my src rpm patch? I notice that after commit `8aa39b2` there are 5 more commits addressing the closed issue [1448](https://github.com/proftpd/proftpd/issues/1448). Some of these commits are on the master branch, however, which is why I chose that particular commit.

Thanks,
Peter

**Comment by Castaglia** [comment]
Interesting.  I've reproduced some bad behavior, using the 1.3.7 branch and OpenSSL 3.x, on an Ubuntu 22.04 Docker image.  I'd backported the fixes to the 1.3.7 branch, but it appears that more might be needed; I'm looking into this now.

**Comment by Castaglia** [comment]
These OpenSSL 3.x issues with the 1.3.7 branch should now be fixed; I'm going to close this ticket.  Please re-open it if you find more issues.  Thanks!
